board.pl

```diff
@@ -53,37 +55,78 @@
         axis(A, direction(W,V))
     ).
 
-% TODO
-% slot(Num, Letter)
+letter_to_row(L, I):-
+    char_code(L, LCode),
+    I is LCode - 64.
 
-% TODO
-% % matches slot's color, depends on board state
-% % slot_color(Board, Slot, Color)
-% slot_color(Board, slot(Num, Letter), color(Color)).
+% matches slot content, depends on board state
+slot_by_index(BoardState, RowIndex, ColIndex, Slot):-
+    nth0(RowIndex, BoardState, RowSlots),
+    nth0(ColIndex, RowSlots, Slot).
 
-% TODO
-% % matches slot that contain a white ball, depends on board state
-% % white_ball(Board, Slot)
-% white_ball(Board, slot(Num, Letter)).
+% matches slot that contain a white ball, depends on board state
+white_ball(BoardState, RowIndex, ColIndex):-
+    slot_by_index(BoardState, RowIndex, ColIndex, 'W').
 
-% TODO
-% % matches slots that contain a black ball, depends on board state
-% % black_ball(Board, Slot)
-% black_ball(Board, slot(Num, Letter)).
+% matches slots that contain a black ball, depends on board state
+black_ball(BoardState, RowIndex, ColIndex):-
+    slot_by_index(BoardState, RowIndex, ColIndex, 'B').
 
-% TODO
-% % matches slots that contain no ball, depends on board state
-% % no_ball(Board, Slot)
-% no_ball(Board, slot(Num, Letter)).
+% matches slots that contain no ball, depends on board state
+no_ball(BoardState, RowIndex, ColIndex):-
+    slot_by_index(BoardState, RowIndex, ColIndex, 0).
 
+% matches border slots, depends on board state (though this remains static throughout the game)
+border(BoardState, RowIndex, ColIndex):-
+    slot_by_index(BoardState, RowIndex, ColIndex, -1).
 
-% TODO
-% % matches all slots of specified row
-% row(slot(Num, Letter), axis(X, Y), Slots).
+% matches all slots of specified horizontal row
+horizontal_row(BoardState, RowIndex, Slots):-
```

Comment:
I see that here you include the '-1' slots while in other predicates you don't (for example in diagonal_row you skip those slots). Maybe you should make it the same in all predicates

Reply:
I do include them also in the diagonal row.
Not sure if I changed it today,
Anyways, we need at least one '-1' in every row we fetch from the board, for the checks